tinygrad/runtime/ops_hip.py

```diff
@@ -55,11 +62,120 @@ class PackageStruct(ctypes.Structure):
+  std::uint32_t block_dim_x,
+  std::uint32_t block_dim_y,
+  std::uint32_t block_dim_z,
+  std::uint32_t shared_mem_bytes,
+  hipStream_t stream,
+"""] + [', '.join([f'{t} {bufnames[i]}' for i,t in buftypes])] + [r"""
+) {
+  hipLaunchKernelGGL(KERNEL_NAME_PLACEHOLDER, dim3(grid_dim_x, grid_dim_y, grid_dim_z), dim3(block_dim_x, block_dim_y, block_dim_z), shared_mem_bytes, stream,
+"""] + [', '.join([f'{bufnames[i]}' for i,t in buftypes])] + [');\nhipStreamSynchronize(stream);\n}']
+
+
+common_kernel_prefix = r"""
+__device__ float4 max(float4 x, float4 y) {
+  return float4(fmax(x.x, y.x), fmax(x.y, y.y), fmax(x.z, y.z), fmax(x.w, y.w));
+}
+
+__device__ float4 pow(float x, float4 y) {
+  return float4(pow(x, y.x), pow(x, y.y), pow(x, y.z), pow(x, y.w));
+}
+
+__device__ float4 pow(float4 x, float4 y) {
+  return float4(pow(x.x, y.x), pow(x.y, y.y), pow(x.z, y.z), pow(x.w, y.w));
+}
+__device__ float4 log2(float4 x) {
+
+  return float4(log2(x.x), log2(x.y), log2(x.z), log2(x.w));
+}
+extern "C" __global__
+"""
+
+common_half_prekernel = r"""
+__device__ float vload_half(size_t offset, const half *p) {
+  return (float)*(p + offset);
+}
+
+__device__ float4 vload_half4(size_t offset, const half * p) {
+  return make_float4((float)*(p + offset *4), (float)*(p + offset *4 + 1), (float)*(p + offset *4 + 2), (float)*(p + offset *4 +3));
+}
+
+__device__ void vstore_half(float data, size_t offset, half *p) {
+  *(p + offset) = (half)data;
+}
+"""
+
 class HIPCodegen(CStyleCodegen):
   lang = CStyleLanguage(
-    kernel_prefix = "#define INFINITY (__builtin_inff())\nextern \"C\" __global__", smem_prefix = "__shared__ ", barrier = "__syncthreads();", float4 = "make_float4",
-    half_prekernel = "",
+    kernel_prefix = r"""
+#include <hip/hip_common.h>
+#include <hip/hip_math_constants.h>
+#define INFINITY (__builtin_inff())
+#define NAN HIP_NAN_F
+""" + common_kernel_prefix,
+    smem_prefix = "__shared__ ", barrier = "__syncthreads();", float4 = "make_float4",
+    half_prekernel = r"""
+#include <hip/hip_fp16.h>
+using half4 = HIP_vector_type<half, 4>;
+""" + common_half_prekernel,
+    uses_vload=True,
+    gid = [f'blockDim.{chr(120+i)}*blockIdx.{chr(120+i)}+threadIdx.{chr(120+i)}' for i in range(3)],
+    lid = [f'threadIdx.{chr(120+i)}' for i in range(3)])
+
+class HIPCodegenCPU(CStyleCodegen):
+  lang = CStyleLanguage(
+    kernel_prefix = r"""
+#include <math.h>
+typedef unsigned char uchar;
+inline float max(float x, float y) {
+  return fmax(x, y);
+}
+""" + common_kernel_prefix,
+    smem_prefix = "__shared__ ", barrier = "__syncthreads();", float4 = "make_float4", uses_vload=True,
+    half_prekernel = r"""
+using half_float::half;
+MAKE_VECTOR_TYPE(half, half)
+""" + common_half_prekernel,
+    need_kernel_launcher = True,
+    build_kernel_launcher = build_kernel_launcher,
     gid = [f'blockDim.{chr(120+i)}*blockIdx.{chr(120+i)}+threadIdx.{chr(120+i)}' for i in range(3)],
     lid = [f'threadIdx.{chr(120+i)}' for i in range(3)])
```

Comment:
Really, the codegen should be the same between HIP and HIP_CPU

Reply:
They can't, since you could see that the required header files, macros and typedefs are different for CPU backend and GPU backend.